src/internal/collection/README.md

```diff
@@ -0,0 +1,103 @@
+
+## Postgres
+
+The postgres implementation is written to mirror the functionality of the etcd implementation rather than to provide an ORM-style table of the protobuf fields inside a database.
+
+### sqlx.Tx
+
+The `sqlx.Tx` can be constructed via `NewSQLTx`.  This will occupy a sql session for its lifetime.  Constructing tehe `sqlx.Tx` will result in a `BEGIN;` at the start, and a `COMMIT;` at the end of the callback, unless an error occurs, in which case a `ROLLBACK;` will be issued.  This will reattempt up to three times in the case of a commit error.
+
+### Tables
+
+A table will be created for each postgres collection.  Construction of the collection will check that the table exists, as well as all relevant triggers and indexes.  If any are missing, they will be created.
+
+The `model` struct exists to mirror the columns in the database for reading results from `sqlx` into a struct:
+ * `CreatedAt` (golang) -> `createdat` (postgres) - the timestamp at which the item was created 
+ * `UpdatedAt` (golang) -> `updatedat` (postgres) - the timestamp at which the item was last modified
+ * `Version` (golang) -> `version` (postgres) - the version of pachyderm which serialized the protobuf
+ * `Key` (golang) -> `key` (postgres) - the primary key for the item
+ * `Proto` (golang) -> `proto` (postgres) - the item's serialized protobuf
+
+ Additionally, columns will be created for each indexed field, following the pattern `idx_<field>` - these are only used for indexing purposes and are never actually read back out by a client, the user can get that information from the protobuf.
+
+### Triggers
+
+Two triggers exist on each postgres collection table, `update_modified_trigger` and `notify_watch_trigger`.
+
+`update_modified_trigger` will modify the row before any write to update the `updatedat` column to the current time.
+
+`notify_watch_trigger` will issue `NOTIFY` commands to one or several channels in case any clients are running watches on the collection.  For table-wide watches, the channel is straightforwardly `pwc_<table>`.  For point-watches, the trigger will also notify one channel for each index on the collection.  The channel name must depend on the value of the field being indexed so that clients do not have to discard O(n) events per interesting event.  Unfortunately, channel names must be valid postgres identifiers which puts several limitations on the channel name, so the value must be hashed.  This still leaves the possibility open for a collision in the channel name, so the client must still check that the event belongs to the watch in question (see [Listeners](#listeners) for details).
```

Comment:
I think there is a bug with pq and names longer than 31 characters.  Not sure if you've run into that.

Reply:
I haven't seen any issues with this, maybe because we use their listener object directly rather than doing this through the `sql` package.